fdbclient/ClientStatusReport.cpp

```diff
@@ -0,0 +1,173 @@
+			serverAddresses.insert(commitProxy.address());
+			commitProxyArr.push_back(commitProxy.address().toString());
+		}
+		statusObj["CommitProxies"] = commitProxyArr;
+
+		// Update health status
+		if (clientInfo.grvProxies.size() == 0 || clientInfo.commitProxies.size() == 0) {
+			healthy = false;
+		}
+	}
+
+	void reportStorageServers() {
+		json_spirit::mArray storageServerArr;
+		for (const auto& [ssid, serverInfo] : cx.server_interf) {
+			json_spirit::mObject serverDescr;
+			serverDescr["SSID"] = ssid.toString();
+			serverDescr["Address"] = serverInfo->interf.address().toString();
+			serverAddresses.insert(serverInfo->interf.address());
+			storageServerArr.push_back(serverDescr);
+		}
+		statusObj["StorageServers"] = storageServerArr;
+	}
+
+	void reportConnections() {
+		json_spirit::mArray connectionArr;
+		for (const auto& addr : serverAddresses) {
+			connectionArr.push_back(connectionStatusReport(addr));
+		}
+		statusObj["Connections"] = connectionArr;
+		statusObj["NumConnectionsFailed"] = numConnectionsFailed;
+
+		// Update health status
+		if (numConnectionsFailed > 0) {
+			healthy = false;
+		}
+	}
+
+	json_spirit::mObject connectionStatusReport(const NetworkAddress& address) {
+		json_spirit::mObject connStatus;
+		connStatus["Address"] = address.toString();
+
+		auto& peers = FlowTransport::transport().getAllPeers();
+		auto peerIter = peers.find(address);
+
+		bool failed = IFailureMonitor::failureMonitor().getState(address).isFailed();
+		if (failed) {
+			connStatus["Status"] = "failed";
+			numConnectionsFailed++;
+		} else if (peerIter == peers.end()) {
+			connStatus["Status"] = "disconnected";
+		} else {
+			connStatus["Status"] = peerIter->second->connected ? "connected" : "connecting";
+		}
+		if (peerIter != peers.end()) {
+			auto peer = peerIter->second;
+			double currTime = now();
+			connStatus["ConnectFailedCount"] = peer->connectFailedCount;
+			connStatus["Compatible"] = peer->compatible;
+			connStatus["LastConnectTime"] = currTime - peer->lastConnectTime;
+			connStatus["PingCount"] = peer->pingLatencies.getPopulationSize();
+			connStatus["PingTimeoutCount"] = peer->timeoutCount;
+			auto lastLoggedTime = (peer->lastLoggedTime > 0.0) ? peer->lastLoggedTime : peer->lastConnectTime;
+			connStatus["BytesSampleTime"] = currTime - lastLoggedTime;
+			connStatus["BytesReceived"] = peer->bytesReceived;
+			connStatus["BytesSent"] = peer->bytesSent;
+			auto protocolVersion = peer->protocolVersion->get();
+			if (protocolVersion.present()) {
+				connStatus["ProtocolVersion"] = format("%llx", protocolVersion.get().version());
+			}
+		}
+
+		return connStatus;
+	}
+
+	DatabaseContext& cx;
+	json_spirit::mObject statusObj;
+	std::set<NetworkAddress> serverAddresses;
+	int numConnectionsFailed;
+	bool healthy;
+};
+
+} // namespace
+
+// Get client-side status information
+Standalone<StringRef> DatabaseContext::getClientStatus() {
+	ClientReportGenerator generator(*this);
```

Comment:
Does it need to be a class if we're immediately calling `generateReport()` on it?

Reply:
It is a class for keeping intermediate state as members instead of passing them around as parameters. At first I had just a set of functions in an anonymous namespace, but then it got messy when passing multiple parameters, especially output parameters like connections and healthy.